File: anshin_theme/public/js/revenue_dashboard.js
// Resource Revenue Management Dashboard - Frontend JavaScript

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June',
                   'July', 'August', 'September', 'October', 'November', 'December'];
const monthNamesShort = ['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN',
                        'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC'];

// Global state
let dashboardData = null;

// Initialize dashboard on page load
function initializeDashboard() {
    const now = new Date();
    const currentMonth = now.getMonth();  // 0-based
    const currentYear = now.getFullYear();

    // Set month selector to current month
    document.getElementById('monthSelect').value = currentMonth;

    // Set year selector to current year
    const yearSelect = document.getElementById('yearSelect');
    const yearOption = Array.from(yearSelect.options).find(opt => opt.value == currentYear);
    if (yearOption) {
        yearSelect.value = currentYear;
    }

    // Load dashboard data
    loadDashboardData(currentMonth, currentYear);
}

// Load dashboard data from API
function loadDashboardData(month, year) {
    // Show loading state
    showLoadingState();

    // Call Frappe API
    frappe.call({
        method: 'anshin_theme.api.revenue_dashboard.get_dashboard_data',
        args: {
            month: month,
            year: year
        },
        callback: function(response) {
            if (response.message) {
                dashboardData = response.message;
                renderDashboard();
                updateLastUpdateTime();
            }
        },
        error: function(error) {
            console.error('Error loading dashboard data:', error);
            frappe.msgprint({
                title: __('Error'),
                indicator: 'red',
                message: __('Failed to load dashboard data. Please try again.')
            });
        }
    });
}

// Show loading state
function showLoadingState() {
    // Add loading spinners or skeleton screens if needed
    console.log('Loading dashboard data...');
}

// Render entire dashboard
function renderDashboard() {
    if (!dashboardData) return;

    renderRevenueSummary();
    renderUtilization();
    renderAllEmployees();
    renderSummaryCards();
    renderOnBench();
    renderExpiringThisMonth();
    renderExpiringNextMonth();
    renderExpiringNext3Months();
    renderTotalAtRisk();
    updateSectionTitles();
}

// Render revenue summary cards
function renderRevenueSummary() {
    const data = dashboardData.revenue_summary;

    // Active Revenue card
    document.getElementById('activeRevenue').textContent = formatCurrency(data.active_revenue);
    document.getElementById('billableCount').textContent = data.billable_count;
    document.getElementById('avgRate').textContent = formatCurrency(data.avg_rate);

    // Net Margin card
    document.getElementById('netMargin').textContent = formatCurrency(data.net_margin);
    document.getElementById('marginPercentage').textContent = `${Math.round(data.margin_percentage)}%`;

    // Bench Cost Impact card
    document.getElementById('benchCost').textContent = formatCurrency(-data.bench_cost);
    document.getElementById('benchResourceCount').textContent = `${data.bench_count} resources without billing`;
    document.getElementById('actionItems').textContent = data.action_items;
    document.getElementById('atRisk').textContent = formatCurrency(data.at_risk);
}

// Render utilization bar
function renderUtilization() {
    const data = dashboardData.utilization;

    // Update utilization bar
    const percentage = Math.round(data.utilization_percentage);
    document.getElementById('utilizationBar').style.width = `${percentage}%`;
    document.getElementById('utilizationBar').textContent = `${percentage}% Utilization`;

    // Update metrics
    document.getElementById('billableMetric').textContent = data.billable;
    document.getElementById('onBenchMetric').textContent = data.on_bench;
    document.getElementById('totalMetric').textContent = data.total;
}

// Render all employees section
function renderAllEmployees() {
    const employees = dashboardData.all_employees;
    const tbody = document.getElementById('allEmployeesTableBody');

    tbody.innerHTML = '';

    employees.forEach(emp => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>
                <div class="resource-name">${emp.employee_name}</div>
            </td>
            <td><span class="resource-id">${emp.employee_id}</span></td>
            <td><span class="cost-value">${formatCurrency(emp.ctc)}</span></td>
            <td><span class="cost-value">${formatCurrency(emp.daily_cost)}</span></td>
            <td>${renderSkills(emp.skills)}</td>
        `;
        tbody.appendChild(row);
    });

    // Update header count
    document.getElementById('totalEmployeesCount').textContent = `${employees.length} Total Employees`;
}

// Render summary cards
function renderSummaryCards() {
    const cards = dashboardData.summary_cards;

    // On Bench Now
    document.getElementById('onBenchNowCount').textContent = cards.on_bench_now.count;
    document.getElementById('onBenchNowAmount').textContent = formatCurrency(cards.on_bench_now.amount) + '/mo';

    // Expiring This Month
    document.getElementById('expiringThisMonthCount').textContent = cards.expiring_this_month.count;
    document.getElementById('expiringThisMonthAmount').textContent = formatCurrency(cards.expiring_this_month.amount) + '/mo';

    // Expiring Next Month
    document.getElementById('expiringNextMonthCount').textContent = cards.expiring_next_month.count;
    document.getElementById('expiringNextMonthAmount').textContent = formatCurrency(cards.expiring_next_month.amount) + '/mo';

    // Expiring Next 3 Months
    document.getElementById('expiringNext3MonthsCount').textContent = cards.expiring_next_3_months.count;
    document.getElementById('expiringNext3MonthsAmount').textContent = formatCurrency(cards.expiring_next_3_months.amount) + ' Total';
}

// Render on bench section
function renderOnBench() {
    const data = dashboardData.on_bench;
    const tbody = document.getElementById('onBenchTableBody');

    tbody.innerHTML = '';

    // Update section header
    document.getElementById('onBenchCount').textContent = `${data.employees.length} Resources`;
    document.getElementById('onBenchLoss').textContent =
        `Daily Loss: ${formatCurrency(data.daily_loss)} | Monthly: ${formatCurrency(data.monthly_cost)}`;

    data.employees.forEach(emp => {
        const row = document.createElement('tr');

        // Format previous client display
        let previousClient = 'New Resource';
        if (emp.last_client) {
            if (emp.contracts_with_last_client > 1) {
                previousClient = `${emp.last_client} (${emp.contracts_with_last_client} contracts)`;
            } else {
                previousClient = `${emp.last_client} (${formatDate(emp.last_contract_end)})`;
            }
        }

        row.innerHTML = `
            <td>
                <div class="resource-name">${emp.employee_name}</div>
                <span class="resource-id">${emp.employee_id}</span>
            </td>
            <td><span class="days-badge ${getDaysBadgeClass(emp.days_on_bench)}">${emp.days_on_bench} days</span></td>
            <td><span class="cost-value">${formatCurrency(emp.daily_cost)}</span></td>
            <td><span class="loss-value">${formatCurrency(emp.total_loss)}</span></td>
            <td>
                <div class="client-info">
                    <span class="client-name">${previousClient}</span>
                </div>
            </td>
            <td>${renderSkills(emp.skills)}</td>
        `;
        tbody.appendChild(row);
    });
}

// Render expiring this month section
function renderExpiringThisMonth() {
    const data = dashboardData.expiring_this_month;
    const tbody = document.getElementById('expiringThisMonthTableBody');

    tbody.innerHTML = '';

    // Update section header
    document.getElementById('expiringThisMonthSectionCount').textContent = `${data.contracts.length} Resources`;
    document.getElementById('expiringThisMonthSectionLoss').textContent =
        `Potential Loss: ${formatCurrency(data.potential_loss_daily)}/day | ${formatCurrency(data.potential_loss_monthly)}/month`;

    data.contracts.forEach(contract => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>
                <div class="resource-name">${contract.employee_name}</div>
                <span class="resource-id">${contract.employee_id}</span>
            </td>
            <td><span class="days-badge ${getDaysBadgeClass(contract.days_left)}">${contract.days_left} days</span></td>
            <td><span class="cost-value">${formatCurrency(contract.daily_cost)}</span></td>
            <td>
                <div class="client-info">
                    <span class="client-name">${contract.current_client}</span>
                </div>
            </td>
            <td>${renderSkills(contract.skills)}</td>
        `;
        tbody.appendChild(row);
    });
}

// Render expiring next month section
function renderExpiringNextMonth() {
    const data = dashboardData.expiring_next_month;
    const tbody = document.getElementById('expiringNextMonthTableBody');

    tbody.innerHTML = '';

    // Update section header
    document.getElementById('expiringNextMonthSectionCount').textContent = `${data.contracts.length} Resources`;
    document.getElementById('expiringNextMonthSectionLoss').textContent =
        `Potential Loss: ${formatCurrency(data.potential_loss_daily)}/day | ${formatCurrency(data.potential_loss_monthly)}/month`;

    data.contracts.forEach(contract => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>
                <div class="resource-name">${contract.employee_name}</div>
                <span class="resource-id">${contract.employee_id}</span>
            </td>
            <td><span class="days-badge ${getDaysBadgeClass(contract.days_left)}">${contract.days_left} days</span></td>
            <td><span class="cost-value">${formatCurrency(contract.daily_cost)}</span></td>
            <td>
                <div class="client-info">
                    <span class="client-name">${contract.current_client}</span>
                </div>
            </td>
            <td>${renderSkills(contract.skills)}</td>
        `;
        tbody.appendChild(row);
    });
}

// Render expiring next 3 months section
function renderExpiringNext3Months() {
    const data = dashboardData.expiring_next_3_months;
    const tbody = document.getElementById('expiringNext3MonthsTableBody');

    tbody.innerHTML = '';

    // Update section header
    document.getElementById('expiringNext3MonthsSectionCount').textContent = `${data.contracts.length} Resources`;
    document.getElementById('expiringNext3MonthsSectionLoss').textContent =
        `Potential Loss: ${formatCurrency(data.potential_loss_daily)}/day | ${formatCurrency(data.potential_loss_monthly)}/month`;

    data.contracts.forEach(contract => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>
                <div class="resource-name">${contract.employee_name}</div>
                <span class="resource-id">${contract.employee_id}</span>
            </td>
            <td><span class="days-badge ${getDaysBadgeClass(contract.days_left)}">${contract.days_left} days</span></td>
            <td><span class="cost-value">${formatCurrency(contract.daily_cost)}</span></td>
            <td>
                <div class="client-info">
                    <span class="client-name">${contract.current_client}</span>
                </div>
            </td>
            <td>${renderSkills(contract.skills)}</td>
        `;
        tbody.appendChild(row);
    });
}

// Render total at risk
function renderTotalAtRisk() {
    document.getElementById('totalAtRisk').textContent = formatCurrency(dashboardData.total_at_risk);
}

// Update section titles based on selected month
function updateSectionTitles() {
    const monthIndex = parseInt(document.getElementById('monthSelect').value);

    // Section 2: This month
    document.getElementById('section2Title').textContent = 'EXPIRING THIS MONTH';
    document.getElementById('summaryCard2Label').textContent = 'Expiring This Month';

    // Section 3: Next month
    const nextMonthIndex = (monthIndex + 1) % 12;
    document.getElementById('section3Title').textContent = monthNames[nextMonthIndex].toUpperCase();
    document.getElementById('summaryCard3Label').textContent = `Expiring in ${monthNames[nextMonthIndex]}`;

    // Section 4: Next 3 months
    const month2Index = (monthIndex + 1) % 12;
    const month4Index = (monthIndex + 3) % 12;
    document.getElementById('section4Title').textContent =
        `${monthNamesShort[month2Index]}-${monthNamesShort[month4Index]}`;
    document.getElementById('summaryCard4Label').textContent =
        `Expiring in Next 3 Months (${monthNamesShort[month2Index]}-${monthNamesShort[month4Index]})`;
}

// Update dashboard when month or year is changed
function updateDashboard() {
    const month = parseInt(document.getElementById('monthSelect').value);
    const year = parseInt(document.getElementById('yearSelect').value);

    loadDashboardData(month, year);
}

// Refresh dashboard
function refreshDashboard() {
    const month = parseInt(document.getElementById('monthSelect').value);
    const year = parseInt(document.getElementById('yearSelect').value);

    loadDashboardData(month, year);

    frappe.show_alert({
        message: __('Dashboard refreshed successfully'),
        indicator: 'green'
    }, 3);
}

// Update last update time
function updateLastUpdateTime() {
    const now = new Date();
    const hours = now.getHours().toString().padStart(2, '0');
    const minutes = now.getMinutes().toString().padStart(2, '0');
    document.getElementById('lastUpdateTime').textContent = `${hours}:${minutes}`;
}

// Toggle section collapse
function toggleSection(sectionId) {
    const section = document.getElementById(sectionId);
    section.classList.toggle('collapsed');
}

// Helper: Format currency (JPY)
function formatCurrency(value) {
    if (value === null || value === undefined) return '¥0';

    const absValue = Math.abs(value);
    const sign = value < 0 ? '-' : '';

    if (absValue >= 1000000) {
        return `${sign}¥${(absValue / 1000000).toFixed(1)}M`;
    } else if (absValue >= 1000) {
        return `${sign}¥${(absValue / 1000).toFixed(0)}K`;
    } else {
        return `${sign}¥${Math.round(absValue).toLocaleString()}`;
    }
}

// Helper: Format date
function formatDate(dateString) {
    if (!dateString) return '';

    const date = new Date(dateString);
    const month = monthNamesShort[date.getMonth()];
    const day = date.getDate();
    const year = date.getFullYear();
    return `${month} ${day}, ${year}`;
}

// Helper: Get badge class for days
function getDaysBadgeClass(days) {
    if (days <= 7) return 'days-critical';
    if (days <= 14) return 'days-urgent';
    if (days <= 30) return 'days-high';
    return 'days-normal';
}

// Helper: Render skills
function renderSkills(skills) {
    if (!skills || skills.length === 0) return '<em>No skills listed</em>';

    return `<div class="skills-list">
        ${skills.map(skill => {
            const experience = skill.no_of_years ? `(${skill.no_of_years}y)` : '';
            return `<span class="skill-item"><span class="skill-name">${skill.skill}</span> <span class="skill-years">${experience}</span></span>`;
        }).join('')}
    </div>`;
}

// Initialize when DOM is ready
if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', initializeDashboard);
} else {
    initializeDashboard();
}

File: anshin_theme/api/revenue_dashboard.py
import frappe
from frappe import _
from datetime import datetime, timedelta
from calendar import monthrange

@frappe.whitelist()
def get_dashboard_data(month=None, year=None):
    """
    Get Resource Revenue Management Dashboard data for specified month/year.

    Args:
        month: Month number (0-11, JavaScript style) or None for current month
        year: Year (e.g., 2025) or None for current year

    Returns:
        Dictionary with all dashboard sections data
    """
    # Default to current month/year if not provided
    today = datetime.now()
    if month is None:
        month = today.month - 1  # Convert to 0-based
    else:
        month = int(month)

    if year is None:
        year = today.year
    else:
        year = int(year)

    # Convert 0-based month to 1-based for calculations
    selected_month = month + 1
    selected_year = year

    # Calculate date ranges
    first_day = datetime(selected_year, selected_month, 1).date()
    last_day = datetime(selected_year, selected_month, monthrange(selected_year, selected_month)[1]).date()

    # Get data for all sections
    data = {
        "revenue_summary": get_revenue_summary(first_day, last_day),
        "utilization": get_utilization_metrics(first_day, last_day),
        "all_employees": get_all_employees(),
        "summary_cards": get_summary_cards(first_day, last_day),
        "on_bench": get_on_bench_employees(first_day, last_day),
        "expiring_this_month": get_expiring_contracts(first_day, last_day, "this_month"),
        "expiring_next_month": get_expiring_contracts(first_day, last_day, "next_month"),
        "expiring_next_3_months": get_expiring_contracts(first_day, last_day, "next_3_months"),
        "total_at_risk": 0,  # Will be calculated from summary cards
        "selected_month": selected_month,
        "selected_year": selected_year
    }

    # Calculate total at risk
    data["total_at_risk"] = (
        data["summary_cards"]["expiring_this_month"]["amount"] +
        data["summary_cards"]["expiring_next_month"]["amount"] +
        data["summary_cards"]["expiring_next_3_months"]["amount"]
    )

    return data


def get_revenue_summary(first_day, last_day):
    """Get revenue summary cards data"""

    # Active Revenue: Sum of custom_unit_price for active contracts in selected month
    active_contracts = frappe.db.sql("""
        SELECT
            SUM(custom_unit_price) as total_revenue,
            COUNT(DISTINCT custom_candidate) as billable_count
        FROM `tabContract`
        WHERE docstatus = 1
          AND custom_candidate IS NOT NULL
          AND start_date <= %(last_day)s
          AND end_date >= %(first_day)s
    """, {"first_day": first_day, "last_day": last_day}, as_dict=True)[0]

    active_revenue = active_contracts.total_revenue or 0
    billable_count = active_contracts.billable_count or 0
    avg_rate = (active_revenue / billable_count) if billable_count > 0 else 0

    # Net Margin: Revenue - Salary Costs
    # Get billable employees and their CTC
    billable_employees = frappe.db.sql("""
        SELECT DISTINCT e.ctc
        FROM `tabEmployee` e
        INNER JOIN `tabContract` c ON c.custom_candidate = e.name
        WHERE c.docstatus = 1
          AND c.start_date <= %(last_day)s
          AND c.end_date >= %(first_day)s
          AND e.status = 'Active'
    """, {"first_day": first_day, "last_day": last_day}, as_dict=True)

    total_salary_costs = sum((emp.ctc or 0) / 12 for emp in billable_employees)
    net_margin = active_revenue - total_salary_costs
    margin_percentage = (net_margin / active_revenue * 100) if active_revenue > 0 else 0

    # Bench Cost Impact
    bench_data = get_bench_cost_impact(first_day, last_day)

    return {
        "active_revenue": active_revenue,
        "billable_count": billable_count,
        "avg_rate": avg_rate,
        "net_margin": net_margin,
        "margin_percentage": margin_percentage,
        "total_salary_costs": total_salary_costs,
        "bench_cost": bench_data["monthly_cost"],
        "bench_count": bench_data["count"],
        "action_items": bench_data["action_items"],
        "at_risk": bench_data["at_risk"]
    }


def get_bench_cost_impact(first_day, last_day):
    """Calculate bench cost impact for selected month"""

    # Get employees on bench (no active contract)
    bench_employees = frappe.db.sql("""
        SELECT
            e.name as employee,
            e.ctc,
            e.ctc / 365 as daily_cost
        FROM `tabEmployee` e
        WHERE e.status = 'Active'
          AND NOT EXISTS (
              SELECT 1 FROM `tabContract` c
              WHERE c.custom_candidate = e.name
                AND c.docstatus = 1
                AND c.start_date <= %(last_day)s
                AND c.end_date >= %(first_day)s
          )
    """, {"first_day": first_day, "last_day": last_day}, as_dict=True)

    # Calculate monthly cost for selected month
    days_in_month = (last_day - first_day).days + 1
    monthly_cost = sum(emp.daily_cost * days_in_month for emp in bench_employees)
    bench_count = len(bench_employees)

    # Action items: on bench + expiring this month + expiring next month
    expiring_this_month = get_expiring_count(first_day, last_day, "this_month")
    expiring_next_month = get_expiring_count(first_day, last_day, "next_month")
    action_items = bench_count + expiring_this_month + expiring_next_month

    # At Risk: Total CTC cost for next 3 months if not placed
    at_risk = calculate_at_risk_amount(first_day, last_day)

    return {
        "monthly_cost": monthly_cost,
        "count": bench_count,
        "action_items": action_items,
        "at_risk": at_risk
    }


def calculate_at_risk_amount(first_day, last_day):
    """Calculate total at risk amount for next 3 months"""

    # Get contracts expiring in next 3 months
    next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
    next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()

    three_months_later = next_month_first + timedelta(days=90)

    expiring_contracts = frappe.db.sql("""
        SELECT
            e.ctc,
            e.ctc / 365 as daily_cost,
            c.end_date
        FROM `tabContract` c
        INNER JOIN `tabEmployee` e ON e.name = c.custom_candidate
        WHERE c.docstatus = 1
          AND c.end_date >= %(next_month_first)s
          AND c.end_date <= %(three_months_later)s
    """, {"next_month_first": next_month_first, "three_months_later": three_months_later}, as_dict=True)

    total_at_risk = 0
    for contract in expiring_contracts:
        # Calculate days from end_date + 1 to end of 3-month window
        days_at_risk = (three_months_later - contract.end_date).days
        total_at_risk += contract.daily_cost * days_at_risk

    return total_at_risk


def get_expiring_count(first_day, last_day, period):
    """Get count of expiring contracts for specified period"""

    if period == "this_month":
        start_date = first_day
        end_date = last_day
    elif period == "next_month":
        next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
        next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()
        start_date = next_month_first
        end_date = datetime(next_month_first.year, next_month_first.month,
                           monthrange(next_month_first.year, next_month_first.month)[1]).date()
    else:
        return 0

    count = frappe.db.count("Contract", filters={
        "docstatus": 1,
        "custom_candidate": ["is", "set"],
        "end_date": ["between", [start_date, end_date]]
    })

    return count


def get_utilization_metrics(first_day, last_day):
    """Get utilization metrics"""

    # Total active employees
    total_employees = frappe.db.count("Employee", filters={"status": "Active"})

    # Billable employees (with active contracts)
    billable_count = frappe.db.sql("""
        SELECT COUNT(DISTINCT custom_candidate)
        FROM `tabContract`
        WHERE docstatus = 1
          AND custom_candidate IS NOT NULL
          AND start_date <= %(last_day)s
          AND end_date >= %(first_day)s
    """, {"first_day": first_day, "last_day": last_day})[0][0] or 0

    # On bench = total - billable
    on_bench_count = total_employees - billable_count

    # Utilization percentage
    utilization = (billable_count / total_employees * 100) if total_employees > 0 else 0

    return {
        "total": total_employees,
        "billable": billable_count,
        "on_bench": on_bench_count,
        "utilization_percentage": utilization
    }


def get_all_employees():
    """Get all active employees with CTC and skills"""

    employees = frappe.db.sql("""
        SELECT
            e.name as employee_id,
            e.employee_name,
            e.ctc,
            e.ctc / 365 as daily_cost
        FROM `tabEmployee` e
        WHERE e.status = 'Active'
        ORDER BY e.employee_name
    """, as_dict=True)

    # Get skills for each employee
    for emp in employees:
        skills = frappe.db.sql("""
            SELECT
                skill,
                proficiency,
                no_of_years
            FROM `tabSkill Child`
            WHERE parent = %(employee)s
              AND parenttype = 'Employee'
            ORDER BY skill
        """, {"employee": emp.employee_id}, as_dict=True)

        emp["skills"] = skills

    return employees


def get_summary_cards(first_day, last_day):
    """Get data for the 4 summary cards"""

    today = datetime.now().date()

    # Card 1: On Bench Now
    on_bench = get_on_bench_summary(first_day, last_day)

    # Card 2: Expiring This Month
    expiring_this = get_expiring_summary(first_day, last_day, "this_month")

    # Card 3: Expiring Next Month
    expiring_next = get_expiring_summary(first_day, last_day, "next_month")

    # Card 4: Expiring in Next 3 Months
    expiring_3months = get_expiring_summary(first_day, last_day, "next_3_months")

    return {
        "on_bench_now": on_bench,
        "expiring_this_month": expiring_this,
        "expiring_next_month": expiring_next,
        "expiring_next_3_months": expiring_3months
    }


def get_on_bench_summary(first_day, last_day):
    """Get on bench summary for card"""

    # Get employees on bench
    bench_employees = frappe.db.sql("""
        SELECT
            e.ctc / 365 as daily_cost
        FROM `tabEmployee` e
        WHERE e.status = 'Active'
          AND NOT EXISTS (
              SELECT 1 FROM `tabContract` c
              WHERE c.custom_candidate = e.name
                AND c.docstatus = 1
                AND c.start_date <= %(last_day)s
                AND c.end_date >= %(first_day)s
          )
    """, {"first_day": first_day, "last_day": last_day}, as_dict=True)

    # Calculate cost for selected month
    days_in_month = (last_day - first_day).days + 1
    monthly_cost = sum(emp.daily_cost * days_in_month for emp in bench_employees)

    return {
        "count": len(bench_employees),
        "amount": monthly_cost
    }


def get_expiring_summary(first_day, last_day, period):
    """Get expiring contracts summary for card"""

    if period == "this_month":
        start_date = first_day
        end_date = last_day
    elif period == "next_month":
        next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
        next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()
        start_date = next_month_first
        end_date = datetime(next_month_first.year, next_month_first.month,
                           monthrange(next_month_first.year, next_month_first.month)[1]).date()
    elif period == "next_3_months":
        next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
        next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()
        start_date = next_month_first
        end_date = next_month_first + timedelta(days=90)
    else:
        return {"count": 0, "amount": 0}

    # Get expiring contracts with employee CTC
    expiring = frappe.db.sql("""
        SELECT
            e.ctc,
            e.ctc / 365 as daily_cost,
            c.end_date
        FROM `tabContract` c
        INNER JOIN `tabEmployee` e ON e.name = c.custom_candidate
        WHERE c.docstatus = 1
          AND c.end_date >= %(start_date)s
          AND c.end_date <= %(end_date)s
    """, {"start_date": start_date, "end_date": end_date}, as_dict=True)

    # Calculate prorated CTC cost
    total_amount = 0
    for contract in expiring:
        if period == "this_month" or period == "next_month":
            # Prorated for remaining days in that month
            days_remaining = (end_date - contract.end_date).days
            total_amount += contract.daily_cost * days_remaining
        else:
            # Total for entire 3-month window
            days_remaining = (end_date - contract.end_date).days
            total_amount += contract.daily_cost * days_remaining

    return {
        "count": len(expiring),
        "amount": total_amount
    }


def get_on_bench_employees(first_day, last_day):
    """Get detailed list of employees on bench"""

    today = datetime.now().date()

    # Get employees on bench
    bench_employees = frappe.db.sql("""
        SELECT
            e.name as employee_id,
            e.employee_name,
            e.ctc,
            e.ctc / 365 as daily_cost
        FROM `tabEmployee` e
        WHERE e.status = 'Active'
          AND NOT EXISTS (
              SELECT 1 FROM `tabContract` c
              WHERE c.custom_candidate = e.name
                AND c.docstatus = 1
                AND c.end_date >= CURDATE()
          )
        ORDER BY e.ctc DESC
    """, as_dict=True)

    for emp in bench_employees:
        # Get when they went on bench (most recent contract end date)
        last_contract = frappe.db.sql("""
            SELECT end_date, party_name
            FROM `tabContract`
            WHERE custom_candidate = %(employee)s
              AND docstatus = 1
            ORDER BY end_date DESC
            LIMIT 1
        """, {"employee": emp.employee_id}, as_dict=True)

        if last_contract:
            bench_start = last_contract[0].end_date + timedelta(days=1)
            days_on_bench = (today - bench_start).days
            emp["days_on_bench"] = days_on_bench
            emp["total_loss"] = emp.daily_cost * days_on_bench
            emp["last_client"] = last_contract[0].party_name
            emp["last_contract_end"] = last_contract[0].end_date

            # Get count of contracts with last client
            contracts_count = frappe.db.count("Contract", filters={
                "custom_candidate": emp.employee_id,
                "party_name": last_contract[0].party_name,
                "docstatus": 1
            })
            emp["contracts_with_last_client"] = contracts_count
        else:
            emp["days_on_bench"] = 0
            emp["total_loss"] = 0
            emp["last_client"] = None
            emp["last_contract_end"] = None
            emp["contracts_with_last_client"] = 0

        # Get skills
        skills = frappe.db.sql("""
            SELECT skill, proficiency, no_of_years
            FROM `tabSkill Child`
            WHERE parent = %(employee)s AND parenttype = 'Employee'
        """, {"employee": emp.employee_id}, as_dict=True)

        emp["skills"] = skills

    # Calculate monthly cost for selected month
    days_in_month = (last_day - first_day).days + 1
    monthly_cost = sum(emp.daily_cost * days_in_month for emp in bench_employees)

    return {
        "employees": bench_employees,
        "monthly_cost": monthly_cost,
        "daily_loss": sum(emp.daily_cost for emp in bench_employees)
    }


def get_expiring_contracts(first_day, last_day, period):
    """Get detailed list of expiring contracts"""

    today = datetime.now().date()

    if period == "this_month":
        start_date = first_day
        end_date = last_day
    elif period == "next_month":
        next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
        next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()
        start_date = next_month_first
        end_date = datetime(next_month_first.year, next_month_first.month,
                           monthrange(next_month_first.year, next_month_first.month)[1]).date()
    elif period == "next_3_months":
        next_month_first = datetime(last_day.year, last_day.month, 1).date() + timedelta(days=32)
        next_month_first = datetime(next_month_first.year, next_month_first.month, 1).date()
        start_date = next_month_first
        end_date = next_month_first + timedelta(days=90)
    else:
        return {"contracts": [], "potential_loss_daily": 0, "potential_loss_monthly": 0}

    # Get expiring contracts
    contracts = frappe.db.sql("""
        SELECT
            e.name as employee_id,
            e.employee_name,
            e.ctc,
            e.ctc / 365 as daily_cost,
            c.end_date,
            c.party_name as current_client
        FROM `tabContract` c
        INNER JOIN `tabEmployee` e ON e.name = c.custom_candidate
        WHERE c.docstatus = 1
          AND c.end_date >= %(start_date)s
          AND c.end_date <= %(end_date)s
        ORDER BY c.end_date
    """, {"start_date": start_date, "end_date": end_date}, as_dict=True)

    for contract in contracts:
        # Days left until contract ends
        contract["days_left"] = (contract.end_date - today).days if contract.end_date >= today else 0

        # Get skills
        skills = frappe.db.sql("""
            SELECT skill, proficiency, no_of_years
            FROM `tabSkill Child`
            WHERE parent = %(employee)s AND parenttype = 'Employee'
        """, {"employee": contract.employee_id}, as_dict=True)

        contract["skills"] = skills

    # Calculate potential loss
    daily_loss = sum(c.daily_cost for c in contracts)
    monthly_loss = sum(c.daily_cost * 30 for c in contracts)  # Approximate

    return {
        "contracts": contracts,
        "potential_loss_daily": daily_loss,
        "potential_loss_monthly": monthly_loss
    }
